core/vm/contracts.go

```diff
@@ -474,88 +432,119 @@ func modExpMultComplexityEip2565(x *big.Int) *big.Int {
+		finalDivisor = 20
+		calcMultComplexity = modExpMultComplexityEip198
+	}
+
+	header := getData(input, 0, 3*32)
+
+	// If base or mod is bigger than uint32, the gas cost will be huge.
+	if !allZero(header[0:28]) || !allZero(header[64:64+28]) {
+		return math.MaxUint64
+	}
+
+	// If exp is bigger than uint32:
+	if !allZero(header[32 : 32+28]) {
+		// Before EIP-7883, 0 multiplication complexity cancels the big exp.
+		if !c.osaka && allZero(header[0:32]) && allZero(header[64:96]) {
+			return minGas
+		}
+		// Otherwise, the gas cost will be huge.
+		return math.MaxUint64
+	}
+
 	var (
-		baseLen = new(big.Int).SetBytes(getData(input, 0, 32))
-		expLen  = new(big.Int).SetBytes(getData(input, 32, 32))
-		modLen  = new(big.Int).SetBytes(getData(input, 64, 32))
+		baseLen = binary.BigEndian.Uint32(header[28:32])
+		expLen  = binary.BigEndian.Uint32(header[32+28 : 64])
+		modLen  = binary.BigEndian.Uint32(header[64+28 : 96])
 	)
 	if len(input) > 96 {
 		input = input[96:]
 	} else {
 		input = input[:0]
 	}
 	// Retrieve the head 32 bytes of exp for the adjusted exponent length
-	var expHead *big.Int
-	if big.NewInt(int64(len(input))).Cmp(baseLen) <= 0 {
-		expHead = new(big.Int)
-	} else {
-		if expLen.Cmp(big32) > 0 {
-			expHead = new(big.Int).SetBytes(getData(input, baseLen.Uint64(), 32))
-		} else {
-			expHead = new(big.Int).SetBytes(getData(input, baseLen.Uint64(), expLen.Uint64()))
-		}
-	}
-	// Calculate the adjusted exponent length
-	var msb int
-	if bitlen := expHead.BitLen(); bitlen > 0 {
-		msb = bitlen - 1
-	}
-	adjExpLen := new(big.Int)
-	if expLen.Cmp(big32) > 0 {
-		adjExpLen.Sub(expLen, big32)
-		if c.osaka { // EIP-7883
-			adjExpLen.Lsh(adjExpLen, 4) // ×16
-		} else {
-			adjExpLen.Lsh(adjExpLen, 3) // ×8
-		}
-	}
-	adjExpLen.Add(adjExpLen, big.NewInt(int64(msb)))
-	adjExpLen = math.BigMax(adjExpLen, common.Big1)
-
-	// Calculate the gas cost of the operation
-	gas := new(big.Int).Set(math.BigMax(modLen, baseLen)) // max_length
-	if c.osaka {
-		// EIP-7883: ModExp Gas Cost Increase
-		gas = modExpMultComplexityEip7883(gas /*max_length */)
-		gas.Mul(gas, adjExpLen)
-		if gas.BitLen() > 64 {
-			return math.MaxUint64
-		}
-
-		return max(500, gas.Uint64())
-	} else if c.eip2565 {
-		// EIP-2565 has three changes compared to EIP-198:
-
-		// 1. Different multiplication complexity
-		gas = modExpMultComplexityEip2565(gas)
-
-		gas.Mul(gas, adjExpLen)
-		// 2. Different divisor (`GQUADDIVISOR`) (3)
-		gas.Div(gas, big3)
-		if gas.BitLen() > 64 {
-			return math.MaxUint64
+	expHeadLen := min(expLen, 32)
+	expOffset := baseLen
+	var expHeadExplicitBytes []byte
+	if expOffset < uint32(len(input)) {
+		expHeadExplicitBytes = input[expOffset : expOffset+min(expHeadLen, uint32(len(input))-expOffset)]
+	}
+	// Compute the exp bit width
+	expBitWidth := uint32(0)
+	for i := 0; i < len(expHeadExplicitBytes); i++ {
+		expByte := expHeadExplicitBytes[i]
+		if expByte != 0 {
+			expTopByteBitWidth := 8 - uint32(bits.LeadingZeros8(expByte))
```

Comment:
I'm wondering whether it'll be more performant to pas the input to 32 bytes and use `uint256.BitLen` instead of going byte by byte

Reply:
In this case this made it slower so I'm not going to change it now. However, I used uint256 for initial lengths and the performance has improved by up to 4% in the full benchmarks.